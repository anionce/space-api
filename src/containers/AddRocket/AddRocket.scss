@import '../../assets/styles/variables';

.form-container {
	display: flex;
	flex-direction: column;
	align-items: center;
	gap: 16px;

	.save-button {
		margin-top: 32px;
	}
	:disabled {
		cursor: not-allowed;
		color: #c6c6c6;
		background-color: #ebebe4;
	}
}
