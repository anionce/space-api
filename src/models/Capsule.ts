export type Capsule = {
	id: string;
	land_landings: number;
	last_update: string;
	launches: string[];
	reuse_count: number;
	serial: string;
	status: string;
	type: string;
	water_landings: number;
};

export enum CapsuleType {
	Dragon10 = 'Dragon 1.0',
	Dragon11 = 'Dragon 1.1',
	Dragon20 = 'Dragon 2.0',
}

export const capsuleFilterTypes: CapsuleType[] = [CapsuleType.Dragon10, CapsuleType.Dragon11, CapsuleType.Dragon20];
