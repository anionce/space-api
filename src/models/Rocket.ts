export type Rocket = {
	id: string | undefined;
	serial: string;
	type: string | undefined;
};

export enum RocketType {
	SuperRocket = 'SuperRocket',
	ExtraRocket = 'ExtraRocket',
	HiperRocket = 'HiperRocket',
}

export const rocketFilterTypes: RocketType[] = [RocketType.SuperRocket, RocketType.HiperRocket, RocketType.ExtraRocket];
