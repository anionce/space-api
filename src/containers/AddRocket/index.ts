import { AddRocket } from './AddRocket';
export { AddRocket };
