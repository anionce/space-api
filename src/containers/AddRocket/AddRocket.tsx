import { useEffect, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { addRocket, updateRocket } from '../../redux/rocketSlice';
import { useLocation, useNavigate, useParams } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { useFilter } from '../../hooks/useFilter';
import { TypeFilter } from '../../components/TypeFilter';
import './AddRocket.scss';
import { Input } from '../../components/Input/Input';
import { rocketFilterTypes } from '../../models/Rocket';
import { Rocket } from '../../models/Rocket';
import React from 'react';

export const AddRocket = () => {
	const dispatch = useAppDispatch();
	const { pathname } = useLocation();
	const { id } = useParams();
	const { t } = useTranslation();
	const navigate = useNavigate();

	const getAllRockets: Rocket[] = useAppSelector(state => state.rocketSlice.rockets);
	const rocket: Rocket | undefined = getAllRockets.find(rocket => rocket.id === id);

	const { type, onTypeChanged, setType } = useFilter(getAllRockets, rocket);
	const [serial, setSerial] = useState<string>(rocket?.serial || '');

	useEffect(() => {
		if (isAddPage) {
			setSerial('');
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [pathname]);

	const canSave: Boolean = Boolean(serial && type);
	const isEditAction: Boolean = Boolean(id && serial && type);
	const isAddPage: Boolean = pathname.includes('add');

	const resetActions = (): void => {
		setSerial('');
		setType('');
		navigate('/rockets');
	};

	const onTitleChanged = (event: React.FormEvent<HTMLInputElement>): void => setSerial(event.currentTarget.value);

	const onSaveRocket = (): void => {
		if (isEditAction) {
			dispatch(updateRocket({ id, serial, type }));
		} else {
			const id = (getAllRockets.length + 1).toString();
			dispatch(addRocket({ id, serial, type }));
		}
		resetActions();
	};

	return (
		<section className='select-control'>
			<Input onTitleChanged={onTitleChanged} serial={serial} />
			<TypeFilter onTypeChanged={onTypeChanged} typeSelected={type} itemFilterTypes={rocketFilterTypes} />
			<form className='form-container'>
				<button className='save-button' type='button' onClick={onSaveRocket} disabled={!canSave}>
					{t('space_api.button.save')}
				</button>
			</form>
		</section>
	);
};
