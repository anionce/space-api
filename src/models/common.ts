import { Capsule, CapsuleType } from './Capsule';
import { Rocket, RocketType } from './Rocket';

export type FilterTypes = RocketType | CapsuleType;

export type Item = Rocket | Capsule;
