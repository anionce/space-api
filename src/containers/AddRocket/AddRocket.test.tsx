import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { screen } from '@testing-library/react';
import { customRender } from '../../assets/tests/utils';
import { AddRocket } from './AddRocket';
import { RocketType } from '../../models/Rocket';

test('renders correcly', () => {
	const { container } = customRender(<AddRocket />);

	expect(container).toMatchSnapshot();
});

test('all elements are displayed', () => {
	customRender(<AddRocket />);

	expect(screen.getByText(/Save rocket/i)).toBeInTheDocument();
});

test('filter types are displayed', () => {
	customRender(<AddRocket />);

	expect(screen.getByText(RocketType.SuperRocket)).toBeInTheDocument();
	expect(screen.getByText(RocketType.ExtraRocket)).toBeInTheDocument();
	expect(screen.getByText(RocketType.HiperRocket)).toBeInTheDocument();
});
