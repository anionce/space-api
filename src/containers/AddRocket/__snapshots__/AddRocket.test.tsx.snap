// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correcly 1`] = `
<div>
  <section
    class="select-control"
  >
    <div
      class="input-container"
    >
      <label
        for="input"
      >
        Search
      </label>
      <input
        id="input"
        maxlength="100"
        value=""
      />
    </div>
    <div
      class="select-container"
    >
      <label
        for="type"
      >
        Type
      </label>
      <select
        id="type"
      >
        <option
          selected=""
        />
        <option
          value="SuperRocket"
        >
          SuperRocket
        </option>
        <option
          value="HiperRocket"
        >
          HiperRocket
        </option>
        <option
          value="ExtraRocket"
        >
          ExtraRocket
        </option>
      </select>
    </div>
    <form
      class="form-container"
    >
      <button
        class="save-button"
        disabled=""
        type="button"
      >
        Save rocket
      </button>
    </form>
  </section>
</div>
`;
